napari/layers/labels/labels.py

```diff
@@ -1113,8 +1117,15 @@ def fill(self, coord, new_label, refresh=True):
 
         # If requested new label doesn't change old label then return
         old_label = np.asarray(self.data[int_coord]).item()
-        if old_label == new_label or (
+        if old_label == new_label:
+            return
+        # If preserve_labels is True and old label is not same as
+        # the previous selected label or background, then also return
+        # this covers the case of using 0 (background) fill to erase
+        # accounting for swap_selected_and_background_labels
```

Comment:
I don't think this comment clarifies enough this interaction; maybe link to the PR so there's more context in case someone in the future needs to touch this code!

Reply:
Thanks! How's the new comment? I tried to lay out the logic more clearly.

Reply:
Great, super clear now!